imports/plugins/included/simple-inventory/server/no-meteor/mutations/updateSimpleInventoryBulk.js

```diff
@@ -0,0 +1,147 @@
+        // This is necessary because all fields are required by the collection schema.
+        $setOnInsert[field] = defaultValues[field];
+      }
+    });
+
+    // if (Object.getOwnPropertyNames($set).length === 1) {
+    //   throw new ReactionError("invalid-param", "You must provide at least one field to update.");
+    // }
+
+    const modifier = { $set, $setOnInsert };
+
+    SimpleInventoryCollectionSchema.validate({
+      $set,
+      $setOnInsert: {
+        ...$setOnInsert,
+        "productConfiguration.productVariantId": productConfiguration.productVariantId,
+        shopId
+      }
+    }, { modifier: true, upsert: true });
+
+    return {
+      updateOne: {
+        filter: {
+          "productConfiguration.productVariantId": productConfiguration.productVariantId,
+          shopId
+        },
+        update: modifier,
+        upsert: true
+      }
+    };
+  });
+
+  let res;
+  try {
+    res = await SimpleInventory.bulkWrite(bulkOperations, { ordered: false });
+  } catch (error) {
+    Logger.error({ ...logCtx, error });
+    error.result.result.writeErrors.forEach((writeError) => {
+      Logger.debug({
+        ...logCtx,
+        error,
+        op: writeError.err.op,
+        reason: writeError.err.errmsg
+      });
+    });
+  }
+
+  const upsertedIds = res.result.upserted.map((each) => each._id);
+
+  // If we inserted, set the "reserved" quantity to what it should be. We could have
+  // put this in the $setOnInsert but then we'd have to do the Orders lookup for
+  // calculating reserved every time, even when only an update happens. It's better
+  // to wait until here when we know whether we inserted.
+  if (res.result.nUpserted > 0) {
+    const upsertedDocsOpData = bulkOperations.filter((op) => upsertedIds.includes(op.updateOne.update.$setOnInsert._id));
+    upsertedDocsOpData.forEach((operationData) => {
+      const arg = {
+        productConfiguration: {
+          productId: operationData.updateOne.update.$setOnInsert["productConfiguration.productId"],
+          productVariantId: operationData.updateOne.filter["productConfiguration.productVariantId"]
+        },
+        shopId: operationData.updateOne.filter.shopId
+      };
+      Logger.debug({ ...logCtx, arg }, "Calling recalculateReservedSimpleInventory");
+      // for bulk version: do not await.
+      context.mutations.recalculateReservedSimpleInventory(context, arg);
+    });
+  }
+
+  if (res.result.nModified > 0) {
+    const modifiedDocsOpData = bulkOperations.filter((op) => !upsertedIds.includes(op.updateOne.update.$setOnInsert._id));
+    const productConfigurations = modifiedDocsOpData.map((operationData) => ({
+      productId: operationData.updateOne.update.$setOnInsert["productConfiguration.productId"]
+    }));
+    Logger.debug({ ...logCtx, productConfigurations }, "Product configurations for afterBulkInventoryUpdate");
+    appEvents.emit("afterBulkInventoryUpdate", { productConfigurations });
```

Comment:
Even though we don't currently use it, keep this consistent with "afterInventoryUpdate" by including the user:

```
{ productConfigurations, updatedBy: context.userId }
```

I'd also add `productVariantId` to the `productConfigurations` in your map, even though we don't use that. Someone else might consume the event and need it.